File: benchmarks/OCaml/safety/objects/n_list.ml
class list = object (s)
  method length = (s#length : int)
end

class nil = object
  inherit list
  method length = 0
end

class cons (hd:int) (tl:list) = object
  inherit list
  method length = 1 + tl#length
end

class main = object (s)
  method mk_n_list n =
    if n = 0 then new nil else new cons n (s#mk_n_list (n-1))
end

let main n =
  assert (n = (new main#mk_n_list n)#length)

[@@@types"
  send_mk_n_list :: (v1 : int) ->
                    (v2 : obj) ->
                    {v3 : obj | _cata_Cata_obj(v3) = v1}
  send_length :: (v1 : obj) ->
                 {v2 : int | v2 = _cata_Cata_obj(v1)}
"]

File: benchmarks/Java/safety/simple/n_list.java
class List extends Object {
    List() {
        super();
    }
    Int length() {
        assert false;
    }
}

class Nil extends List {
    Nil() {
        super();
    }
    Int length() {
        return 0;
    }
}

class Cons extends List {
    Int hd;
    List tl;
    Cons(Int hd, List tl) {
        super();
        this.hd = hd;
        this.tl = tl;
    }
    Int length() {
        return 1 + this.tl.length();
    }
}

class Main extends Object {
    Main() { super(); }
    List mk_n_list(Int n) {
        return if n = 0 then new Nil() else new Cons(n, this.mk_n_list(n-1));
    }
}

main(){
    assert (3 = new Main().mk_n_list(3).length());
}
